source/token_provider_sso_profile.c

```diff
@@ -0,0 +1,184 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/auth/credentials.h>
+
+#include <aws/auth/private/aws_profile.h>
+#include <aws/auth/private/credentials_utils.h>
+#include <aws/auth/private/sso_token_providers.h>
+#include <aws/auth/private/sso_token_utils.h>
+#include <aws/common/clock.h>
+
+#ifdef _MSC_VER
+/* allow non-constant declared initializers. */
+#    pragma warning(disable : 4204)
+#endif
+
+/*
+ * sso-token profile provider implementation
+ */
+struct aws_token_provider_profile_impl {
+    struct aws_string *token_file_path;
+
+    aws_io_clock_fn *system_clock_fn;
+};
+
+static int s_token_provider_profile_get_token_async(
+    struct aws_credentials_provider *provider,
+    aws_on_get_credentials_callback_fn callback,
+    void *user_data) {
+    struct aws_token_provider_profile_impl *impl = provider->impl;
+
+    struct aws_sso_token *sso_token = NULL;
+    struct aws_credentials *credentials = NULL;
+    int result = AWS_OP_ERR;
+    sso_token = aws_sso_token_new_from_file(provider->allocator, impl->token_file_path);
+    if (!sso_token) {
+        AWS_LOGF_ERROR(AWS_LS_AUTH_CREDENTIALS_PROVIDER, "(id=%p) unable to read file.", (void *)provider);
+        goto done;
+    }
+
+    /* check token expiration. */
+    uint64_t now_ns = UINT64_MAX;
+    if (impl->system_clock_fn(&now_ns) != AWS_OP_SUCCESS) {
+        goto done;
+    }
+    uint64_t now_seconds = aws_timestamp_convert(now_ns, AWS_TIMESTAMP_NANOS, AWS_TIMESTAMP_SECS, NULL);
+
+    if (aws_date_time_as_epoch_secs(&sso_token->expiration) - now_seconds < 0) {
+        AWS_LOGF_ERROR(AWS_LS_AUTH_CREDENTIALS_PROVIDER, "(id=%p) cached token is expired.", (void *)provider);
+        aws_raise_error(AWS_AUTH_SSO_TOKEN_EXPIRED);
+        goto done;
+    }
+
+    credentials = aws_credentials_new_token(
+        provider->allocator,
+        aws_byte_cursor_from_string(sso_token->token),
+        (uint64_t)aws_date_time_as_epoch_secs(&sso_token->expiration));
+    if (!credentials) {
+        AWS_LOGF_ERROR(AWS_LS_AUTH_CREDENTIALS_PROVIDER, "(id=%p) Unable to construct credentials.", (void *)provider);
+        goto done;
+    }
+
+    callback(credentials, AWS_OP_SUCCESS, user_data);
+    result = AWS_OP_SUCCESS;
+
+done:
+    aws_sso_token_destroy(sso_token);
```

Comment:
make sure to invoke callback on failure as well

The callback should be invoked for both failure and success.

Reply:
The result will be AWS_OP_ERR, so we should not invoke the callback.